main.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"database/sql"
+	"flag"
+	"fmt"
+	"html/template"
+	"log"
+	"os"
+	"time"
+
+	"github.com/caarlos0/env"
+	"github.com/ilbambino/csvomatic/parameters"
+	"github.com/ilbambino/csvomatic/stats"
+	"github.com/ilbambino/csvomatic/worker"
+	_ "github.com/lib/pq"
+)
+
+var numberWorkers int
+var csvFile string
+
+func setupArgs() {
+	flag.IntVar(&numberWorkers, "workers", 2, "Number of workers")
+	flag.StringVar(&csvFile, "input", "query_params.csv", "Path to csv with query arguments")
+	flag.Parse()
+
+}
+
+// postgresConf holds the configuration to connect to the DB
+type postgresConf struct {
+	Port     int    `env:"DB_PORT" envDefault:"5432"`
+	Host     string `env:"DB_HOST" envDefault:"localhost"`
+	User     string `env:"DB_USER" envDefault:"postgres"`
+	Password string `env:"DB_PASSWORD" envDefault:"password"`
+	Database string `env:"DB_NAME" envDefault:"homework"`
+}
+
+// getDBConnectionFromEnv constructs a Postgres connection string from env vars or from defaults
+func getDBConnectionFromEnv() (string, error) {
+
+	cfg := postgresConf{}
+	err := env.Parse(&cfg)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=disable", cfg.User, cfg.Password, cfg.Host, cfg.Port, cfg.Database), nil
+}
+
+func main() {
+
+	commandStart := time.Now()
+	setupArgs()
+
+	connStr, err := getDBConnectionFromEnv()
+	if err != nil {
+		log.Fatal(err)
+	}
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	fileReader, err := os.Open(csvFile)
```

Comment:
what if you wanted to the read file concurrently with querying? what are the pros / cons of that?

(you can probably just describe how it'd change the design)

Reply:
One of my assumptions was that the input file wasn't too big, and hence it made sense to read it at once and then start the queries.

The biggest benefit in case of do both concurrently is better use of the resources:

- Lots of memory you might need to load it (if big)
- Time it takes to load and process you could start doing queries

The downside is that it complicates things: 

- Handle errors in a stream. What do you do if an error in the middle of the stream? (this is the easier to solve, just take a stance and what to do).
- Code changes (to stream the jobs), and the reduce phase.
- Stats on a stream

The reading and streaming is almost ready to work like that. We have this to queue jobs to the workers
```golang
	pool.Queue(worker.Job{QueryParams: query})
```
And instead of doing it after the `ReadFromCSV`, we could merge those into one and put the jobs directly into the queue instead of putting them into memory.

Where it would get a little more complicated is in the "reduce" phase. Right now it is done once all queries have been done and it just accumulates the results (that's the reason the results channel needs to be as big as the possible results). 

But if we do it in a streaming way, the "reduce" phase should be done in a goroutine concurrently with the processing, as you don't know how many results we are going to have (and at some point the channel will fill up).

Lastly, most of the stats being calculated can be easily done in a stream, you just need to keep accumulators or registers (min, max, avg, total). But the median is a little more complicated case. To do a median you need to have all the items when calculating it, which in a stream you don't. You could use any of the different quantile stream approximation algorithms to do it.

So if the median doesn't need to be exact, it is perfectly doable with some changes. It just complicates things, and I think the size of the input should be really big to consider it.

Reply:
Sounds good, thanks for the thought experiment!